src/mapper/pkg/resolvers/helpers.go

```diff
@@ -1,47 +1,86 @@
 package resolvers
 
 import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
 	"github.com/amit7itz/goset"
+	"github.com/otterize/network-mapper/src/mapper/pkg/config"
 	"github.com/otterize/network-mapper/src/mapper/pkg/graph/model"
+	"github.com/otterize/network-mapper/src/shared/kubeutils"
 	"github.com/samber/lo"
+	"github.com/spf13/viper"
+	"io"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sort"
 	"sync"
 )
 
+type intentsHolderStore map[model.OtterizeServiceIdentity]map[string]*goset.Set[model.OtterizeServiceIdentity]
+
+func (s intentsHolderStore) MarshalJSON() ([]byte, error) {
+	// OtterizeServiceIdentity cannot be serialized as map key in JSON, because it is represented as a map itself
+	// therefore, we serialize the store as a slice of [Key, Value] "tuples"
+	return json.Marshal(lo.ToPairs(s))
+}
+
+func (s intentsHolderStore) UnmarshalJSON(b []byte) error {
+	var pairs []lo.Entry[model.OtterizeServiceIdentity, map[string]*goset.Set[model.OtterizeServiceIdentity]]
+	err := json.Unmarshal(b, &pairs)
+	if err != nil {
+		return err
+	}
+	for _, pair := range pairs {
+		s[pair.Key] = pair.Value
+	}
+	return nil
+}
+
 type intentsHolder struct {
-	store map[model.OtterizeServiceIdentity]map[string][]model.OtterizeServiceIdentity
-	lock  sync.Mutex
+	store  intentsHolderStore
+	lock   sync.Mutex
+	client client.Client
 }
 
-func NewIntentsHolder() *intentsHolder {
+func NewIntentsHolder(client client.Client) *intentsHolder {
 	return &intentsHolder{
-		store: make(map[model.OtterizeServiceIdentity]map[string][]model.OtterizeServiceIdentity),
-		lock:  sync.Mutex{},
+		store:  make(intentsHolderStore),
+		lock:   sync.Mutex{},
+		client: client,
 	}
 }
 
 func (i *intentsHolder) Reset() {
 	i.lock.Lock()
 	defer i.lock.Unlock()
 
-	i.store = make(map[model.OtterizeServiceIdentity]map[string][]model.OtterizeServiceIdentity)
+	i.store = make(intentsHolderStore)
 }
 
 func (i *intentsHolder) AddIntent(srcService model.OtterizeServiceIdentity, dstService model.OtterizeServiceIdentity) {
 	i.lock.Lock()
 	defer i.lock.Unlock()
 	intentMap, ok := i.store[srcService]
 	if !ok {
-		intentMap = make(map[string][]model.OtterizeServiceIdentity)
+		intentMap = make(map[string]*goset.Set[model.OtterizeServiceIdentity])
```

Comment:
consider declaring a type for `map[string]*goset.Set[model.OtterizeServiceIdentity]` and using it everywhere to make this part of the code clearer

Reply:
Instead of doing this, I really simplified the intents store, because it was complicated for no reason